src/utils/scrap_data.ts

```diff
@@ -0,0 +1,176 @@
+import puppeteer, { Browser, Page } from 'puppeteer';
+import { VIVINO_GRAPES_API, VIVINO_SEARCH_URL, VIVINO_EXPLORE_API, WINES_PER_PAGE, VIVINO_WINE_STYLES_API, VIVINO_FOODS_API } from '../const/const';
+import { REGIONS } from '../const/regions';
+import { WINE_TYPES } from '../const/wine_types';
+import { WineParams } from '../interfaces/params.interface';
+import { EXPLORE_API_PARAMS } from '../enums/explore_api_params.enum';
+import { Grape, Wine, WineStyle, Food } from '../interfaces/wine.interface';
+
+class ScrapData {
+  grapes: Grape[] = [];
+  wineStyles: WineStyle[] = [];
+  foods: Food[];
+
+  appendParam = (param: string, queryKey: string, list: any = []) => {
+    if (!param) {
+      return '';
+    }
+    let queryString = '';
+    const paramNames = param.split(',');
+    paramNames.forEach(name => {
+      if (list.length == 0) {
+        queryString += `&${queryKey}=${name}`;
+      } else {
+        const obj = list.find((g: any) => g.name.toUpperCase().indexOf(name.toUpperCase()) > -1);
+        if (obj) {
+          queryString += `&${queryKey}=${obj.id}`;
+        }
+      }
+    });
+    return queryString;
+  };
+
+  getVivinoSearchUrl = (pageIndex: number, wineParams: WineParams): string => {
+    wineParams.minPrice = isNaN(wineParams.minPrice) ? 0 : wineParams.minPrice;
+    wineParams.maxPrice = isNaN(wineParams.maxPrice) ? 99999999 : wineParams.maxPrice;
+    wineParams.minRating = isNaN(wineParams.minRating) ? 1 : wineParams.minRating;
+    wineParams.orderBy = !wineParams.orderBy ? 'ratings_average' : wineParams.orderBy;
+
+    let url = `${VIVINO_EXPLORE_API}?${EXPLORE_API_PARAMS.country_code}=US&${EXPLORE_API_PARAMS.currency_code}=USD&${EXPLORE_API_PARAMS.grape_filter}=varietal`;
+    url += `&${EXPLORE_API_PARAMS.order_by}=${wineParams.orderBy}&${EXPLORE_API_PARAMS.order}=desc&${EXPLORE_API_PARAMS.page}=${pageIndex}`;
+    url += this.appendParam(wineParams.grapes, EXPLORE_API_PARAMS.grape_ids, this.grapes);
+    url += this.appendParam(wineParams.wineStyles, EXPLORE_API_PARAMS.wine_style_ids, this.wineStyles);
+    url += this.appendParam(wineParams.foodPairings, EXPLORE_API_PARAMS.food_ids, this.foods);
+    url += `&${EXPLORE_API_PARAMS.min_rating}=${wineParams.minRating}`;
+    url += `&${EXPLORE_API_PARAMS.price_range_min}=${wineParams.minPrice}`;
+    url += `&${EXPLORE_API_PARAMS.price_range_max}=${wineParams.maxPrice}`;
+    if (wineParams.countryCodes) {
+      url += this.appendParam(wineParams.countryCodes, EXPLORE_API_PARAMS.country_codes);
+    }
+    if (wineParams.regions) {
+      url += this.appendParam(wineParams.regions, EXPLORE_API_PARAMS.region_ids, REGIONS);
+    }
+    if (wineParams.wineTypes) {
+      url += this.appendParam(wineParams.wineTypes, EXPLORE_API_PARAMS.wine_type_ids, WINE_TYPES);
+    }
+
+    return url;
+  };
+
+  mapWinesToWineObjects(response: any = []): Wine[] {
+    return response.map((wine: any) => {
+      return {
+        title: wine?.vintage?.name,
```

Comment:
If I would ask you to add the validation of the API response how would you do it?

Reply:
I will add the validation at the starting of the function like given below.
```
mapWinesToWineObjects(response: any = []): Wine[] {
    if (!response || !isArray(response)) {
      return [];
    }
    return response.map((wine: any) => {
      return {
        title: wine?.vintage?.name,
        vintage: wine?.vintage?.year,
        averagePrice: wine?.price?.amount,
        currency: wine?.price?.currency?.code,
        averageRating: wine?.vintage?.statistics?.ratings_average,
        numberOfRatings: wine?.vintage?.statistics?.ratings_count,
        imageUrl: wine?.vintage?.image?.location,
        region: wine?.vintage?.wine?.region?.name,
        country: wine?.vintage?.wine?.region?.country?.name,
        bottleVolume: wine?.price?.bottle_type?.volume_ml,
        discountPercent: wine?.price?.discount_percent,
      };
    });
  }
```